xcdat/temporal.py

```diff
@@ -919,43 +919,40 @@ def _group_average(self, data_var: xr.DataArray) -> xr.DataArray:
 
         if self._weighted:
             self._weights = self._get_weights()
-            # grab metadata
-            dv_attrs = dv.attrs
-            dv_name = dv.name
-            # grab metadata (populated after calling ._group_data())
-            lt_attrs = self._labeled_time.attrs
-            lt_encoding = self._labeled_time.encoding
-            # weight the data variable
+            # Weight the data variable.
             dv *= self._weights
-            # cast the weights to match the dv.shape / dims
-            weights, x = xr.broadcast(self._weights, dv)
-            # ensure missing data receives no weight
+
+            # Ensure missing data (`np.nan`) receives no weight (zero). To
+            # achieve this, first broadcast the one-dimensional (temporal
+            # dimension) shape of the `weights` DataArray to the
+            # multi-dimensional shape of its corresponding data variable.
+            weights, _ = xr.broadcast(self._weights, dv)
             weights = xr.where(np.isnan(dv), 0.0, weights)
-            # perform weighted average
-            dv = self._group_data(dv).sum() / self._group_data(weights).sum()
-            # add dv attributes
-            dv.attrs = dv_attrs
-            dv.name = dv_name
+
+            # Perform weighted average using the formula
+            # WA = sum(data*weights) / sum(weights). The denominator must be
+            # included to take into account zero weight for missing data.
+            with xr.set_options(keep_attrs=True):
+                dv = self._group_data(dv).sum() / self._group_data(weights).sum()
```

Comment:
`xr.set_options(keep_attrs=True)` will preserve the attrs for the `dv`. It does not restore the attrs for the time dimension, so we manually restore those in lines of code below.